crates/sui-types/src/temporary_store.rs

```diff
@@ -1054,15 +1039,10 @@ impl<S: ObjectStore> TemporaryStore<S> {
                         old_obj.storage_rebate
                     } else {
                         // else, this is a dynamic field, not an input object
-                        if let Ok(Some(old_obj)) = self.store.get_object(object_id) {
-                            let expected_version = object.version();
-                            if old_obj.version() != expected_version {
-                                // not a lot we can do safely and under this condition everything is broken
-                                panic!(
-                                    "Expected to find old object with version {expected_version}, found {}",
-                                    old_obj.version(),
-                                );
-                            }
+                        let expected_version = object.version();
+                        if let Ok(Some(old_obj)) =
+                            self.store.get_object_by_key(object_id, expected_version)
```

Comment:
Same as above, let's keep the panic here.

Reply:
isn't that what is happening? this is panicking on `None`.
I have gated the code so that previous version will not change behavior (panic on version mismatch), whereas this is just looking for a specific version and panicking if `None`